test/QueryTest.cpp

```diff
@@ -0,0 +1,116 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include <znc/Query.h>
+#include <znc/User.h>
+#include <znc/Client.h>
+#include <znc/IRCNetwork.h>
+#include <znc/ZNCString.h>
+#include <znc/ZNCDebug.h>
+#include <znc/znc.h>
+
+class QueryTest : public ::testing::Test {
+protected:
+	void SetUp() { CZNC::CreateInstance(); }
+	void TearDown() { CZNC::DestroyInstance(); }
+};
+
+TEST_F(QueryTest, Name) {
+	CUser user("user");
+	CIRCNetwork network(&user, "network");
+
+	CQuery query("query", &network);
+	EXPECT_EQ("query", query.GetName());
+}
+
+TEST_F(QueryTest, AddClearBuffer) {
+	CUser user("user");
+	CIRCNetwork network(&user, "network");
+
+	CQuery query("query", &network);
+	EXPECT_TRUE(query.GetBuffer().IsEmpty());
+	query.AddBuffer("foo");
+	EXPECT_EQ(1, query.GetBuffer().Size());
+	query.AddBuffer("bar");
+	EXPECT_EQ(2, query.GetBuffer().Size());
+	query.ClearBuffer();
+	EXPECT_TRUE(query.GetBuffer().IsEmpty());
+}
+
+TEST_F(QueryTest, BufferSize) {
+	CUser user("user");
+	CIRCNetwork network(&user, "network");
+
+	CQuery query("query", &network);
+	EXPECT_EQ(50, user.GetQueryBufferSize());
+	EXPECT_EQ(50, query.GetBufferCount());
+
+	EXPECT_EQ(500, CZNC::Get().GetMaxBufferSize());
+	EXPECT_FALSE(query.SetBufferCount(1000, false));
+	EXPECT_EQ(50, query.GetBufferCount());
+	EXPECT_TRUE(query.SetBufferCount(500, false));
+	EXPECT_EQ(500, query.GetBufferCount());
+	EXPECT_TRUE(query.SetBufferCount(1000, true));
+	EXPECT_EQ(1000, query.GetBufferCount());
+}
+
+class TestClient : public CClient {
+public:
+	bool Write(const CString& sData) override {
+		lines.push_back(sData);
+		return true;
+	}
+	VCString lines;
+};
+
+TEST_F(QueryTest, SendBuffer) {
+	CUser user("user");
+	CIRCNetwork network(&user, "network");
+	CDebug::SetDebug(false);
+
+	TestClient client;
+	client.SetNick("me");
+	client.AcceptLogin(user);
+	client.lines.clear();
+
+	CQuery query("query", &network);
+	query.AddBuffer(":sender PRIVMSG {target} :{text}", "a message");
+	query.AddBuffer(":me PRIVMSG someone :{text}", "a self-message");
+	query.AddBuffer(":sender NOTICE #znc :{text}", "a notice");
+
+	client.lines.clear();
+	query.SendBuffer(&client);
+	EXPECT_EQ(2, client.lines.size());
+	EXPECT_TRUE(client.lines[0].WildCmp(":sender PRIVMSG me :[??:??:??] a message\r\n")) << client.lines[0];
+	EXPECT_TRUE(client.lines[1].WildCmp(":sender NOTICE #znc :[??:??:??] a notice\r\n")) << client.lines[1];
+
+	client.lines.clear();
+	user.SetTimestampPrepend(false);
+	query.SendBuffer(&client);
+	EXPECT_EQ(2, client.lines.size());
+	EXPECT_EQ(":sender PRIVMSG me :a message\r\n", client.lines[0]);
+	EXPECT_EQ(":sender NOTICE #znc :a notice\r\n", client.lines[1]);
```

Comment:
```c++
using testing::ElementsAre;

EXPECT_THAT(client.lines, ElementsAre("...", "..."));
```
This includes "size" and both elements, in this order.

Variant with `.WildCmp` is less trivial, but it can be replaced with:
```c++
using testing::MatchesRegex;

EXPECT_THAT(client.lines, ElementsAre(MatchesRegex(R"(:sender PRIVMSG me :\[\d\d:\d\d:\d\d\] ...)"), MatchesRegex("...")));
```
Raw string literal will come to help with escaping of `\`

https://code.google.com/p/googlemock/wiki/CheatSheet#String_Matchers for reference.

PS. the advice for `<<` was actually useless :-P